### DISCUSSION TITLE
AI-Deck: CPX with JTAG flashing?

### DISCUSSION THREAD

**Discussion by LorenzoLamberti94** [bitcraze/discussions#956] [Q&A]
Hi !
I am trying out the new CPX protocol on the ai deck
flashing goes smooth via radio, all the examples work fine :) it's great!

However, for R&D, flashing over the air and rebooting the drone every time is very time consuming.

- I was able to flash via jtag, with "make clean all flash".
- however, simply running the code with a "make all run" gets stuck --> im trying the wifi example now

Probably, i am guessing that this way i am breaking the bootloader, or some handshaking with the crazyflie doesnt happen (i guess happens at startup only)

Do you have any suggestions ? do you think it is possible to run the code through jtag ?
thanks  a lot !

**Comment by knmcguire** [comment]
hi Lorenzo!

Yeah you have a point... I don't think we have tried the `make all run` for a while now and I wouldn't know an good alternative to do debugging. Let us try some things out and let me ask around and hopefully get back to you shortly

**Comment by evoggy** [comment]
Hi!

I've looked though the code a bit to refresh my memory and from what I can see (and remember) it's ok to re-start the SPI link (between the ESP32 and the GAP8) but probably not during communication. Note that the GAP8 will start in bootloader mode by default (if the bootloader is flashed) and the ESP32 will send a command via CPX to jump to the firmware ([aideck-esp-firmware/blob/main/main/system.c#L53](https://github.com/bitcraze/aideck-esp-firmware/blob/main/main/system.c#L53)). So if you're not using the bootloader on the GAP8 you need to receive and ignore this message.

Unfortunately the story is different for the UART link between the STM32 and the ESP32. The system supports either of the MCUs starting up first, but then they will sync (by sending/receiving 0xFF from eachother). So restarting the AI deck will break the CF<->AI deck communication, you will have to also restart the CF. Below are the link to the sync code in the ESP32 and the STM32.

[aideck-esp-firmware/blob/main/main/uart_transport.c#L113](https://github.com/bitcraze/aideck-esp-firmware/blob/main/main/uart_transport.c#L113)

[aideck-esp-firmware/blob/main/main/uart_transport.c#L113](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/cpx/cpx_uart_transport.c#L144)
